File: hushh-webapp/ios/App/App/Plugins/HushhContactsPlugin.swift
import Foundation
import Capacitor
import Contacts
import UIKit

/**
 * HushhContactsPlugin - read-only contact lookup for Connect matching.
 *
 * Contacts are returned to the web layer for in-memory hashing only. The web
 * layer sends hashes to the backend and does not persist raw contact records.
 *
 * Enumeration always runs on a background queue. CNContactStore walks the
 * store synchronously, so a few thousand contacts freeze the UI if that work
 * lands on the main thread.
 */
@objc(HushhContactsPlugin)
public class HushhContactsPlugin: CAPPlugin, CAPBridgedPlugin {

    public let identifier = "HushhContactsPlugin"
    public let jsName = "HushhContacts"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "getPermissionState", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "requestPermission", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "openAppSettings", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "readContacts", returnType: CAPPluginReturnPromise)
    ]

    private let store = CNContactStore()
    private let readQueue = DispatchQueue(
        label: "ai.hushh.contacts.read",
        qos: .userInitiated
    )

    private static let defaultLimit = 5000
    private static let maxLimit = 10000

    @objc func getPermissionState(_ call: CAPPluginCall) {
        call.resolve(["state": permissionState()])
    }

    @objc func requestPermission(_ call: CAPPluginCall) {
        let status = CNContactStore.authorizationStatus(for: .contacts)
        guard status == .notDetermined else {
            call.resolve(["state": permissionState()])
            return
        }
        store.requestAccess(for: .contacts) { [weak self] _, _ in
            call.resolve(["state": self?.permissionState() ?? "unavailable"])
        }
    }

    @objc func openAppSettings(_ call: CAPPluginCall) {
        guard let url = URL(string: UIApplication.openSettingsURLString) else {
            call.resolve(["opened": false])
            return
        }
        DispatchQueue.main.async {
            UIApplication.shared.open(url, options: [:]) { opened in
                call.resolve(["opened": opened])
            }
        }
    }

    @objc func readContacts(_ call: CAPPluginCall) {
        let limit = max(1, min(call.getInt("limit") ?? Self.defaultLimit, Self.maxLimit))
        let status = CNContactStore.authorizationStatus(for: .contacts)

        if status == .notDetermined {
            store.requestAccess(for: .contacts) { [weak self] granted, error in
                guard let self else { return }
                if let error = error {
                    call.reject("Contacts permission failed: \(error.localizedDescription)")
                    return
                }
                guard granted else {
                    call.reject("Contacts permission was not granted.")
                    return
                }
                // Stay off the main thread: this is the enumeration path.
                self.readQueue.async { self.resolveContacts(call, limit: limit) }
            }
            return
        }

        guard isReadableStatus(status) else {
            call.reject("Contacts permission was not granted.")
            return
        }
        readQueue.async { [weak self] in
            self?.resolveContacts(call, limit: limit)
        }
    }

    /// iOS 18 added partial access, where the read succeeds but only returns a
    /// user-picked subset. It is readable, but callers must not treat an empty
    /// result as "nobody matched".
    private func isLimitedStatus(_ status: CNAuthorizationStatus) -> Bool {
        if #available(iOS 18.0, *) {
            return status == .limited
        }
        return false
    }

    private func isReadableStatus(_ status: CNAuthorizationStatus) -> Bool {
        return status == .authorized || isLimitedStatus(status)
    }

    private func permissionState() -> String {
        let status = CNContactStore.authorizationStatus(for: .contacts)
        if isLimitedStatus(status) {
            return "limited"
        }
        switch status {
        case .authorized:
            return "granted"
        case .notDetermined:
            return "prompt"
        case .denied:
            return "denied"
        case .restricted:
            return "restricted"
        @unknown default:
            return "unavailable"
        }
    }

    /// Region used to read bare national numbers such as `9876543210`.
    private func deviceRegion() -> String? {
        let region: String?
        if #available(iOS 16.0, *) {
            region = Locale.current.region?.identifier
        } else {
            region = Locale.current.regionCode
        }
        guard let region, region.count == 2 else { return nil }
        return region.uppercased()
    }

    private func resolveContacts(_ call: CAPPluginCall, limit: Int) {
        let keys: [CNKeyDescriptor] = [
            CNContactIdentifierKey as CNKeyDescriptor,
            CNContactGivenNameKey as CNKeyDescriptor,
            CNContactFamilyNameKey as CNKeyDescriptor,
            CNContactOrganizationNameKey as CNKeyDescriptor,
            CNContactPhoneNumbersKey as CNKeyDescriptor
        ]
        let request = CNContactFetchRequest(keysToFetch: keys)
        // Unified contacts already merge linked cards; sorting is pure overhead
        // for a hash-matching pass.
        request.unifyResults = true
        request.sortOrder = .none

        var contacts: [[String: Any]] = []
        var totalAvailable = 0

        do {
            try store.enumerateContacts(with: request) { contact, _ in
                let phoneNumbers = contact.phoneNumbers
                    .map { $0.value.stringValue.trimmingCharacters(in: .whitespacesAndNewlines) }
                    .filter { !$0.isEmpty }
                if phoneNumbers.isEmpty {
                    return
                }
                totalAvailable += 1
                // Keep walking past the limit so `totalAvailable` stays honest
                // and the web layer can tell the user their book was cut short.
                guard contacts.count < limit else { return }

                let nameParts = [
                    contact.givenName.trimmingCharacters(in: .whitespacesAndNewlines),
                    contact.familyName.trimmingCharacters(in: .whitespacesAndNewlines)
                ].filter { !$0.isEmpty }
                let displayName = nameParts.joined(separator: " ")
                contacts.append([
                    "id": contact.identifier,
                    "displayName": displayName.isEmpty ? contact.organizationName : displayName,
                    "phoneNumbers": phoneNumbers
                ])
            }

            var payload: [String: Any] = [
                "contacts": contacts,
                "sourcePlatform": "ios",
                "limited": isLimitedStatus(CNContactStore.authorizationStatus(for: .contacts)),
                "truncated": totalAvailable > contacts.count,
                "totalAvailable": totalAvailable
            ]
            if let region = deviceRegion() {
                payload["defaultRegion"] = region
            }
            call.resolve(payload)
        } catch {
            call.reject("Contacts could not be read: \(error.localizedDescription)")
        }
    }
}

File: hushh-webapp/android/app/src/main/java/com/hussh/app/plugins/HushhContacts/HushhContactsPlugin.kt
package com.hussh.app.plugins.HushhContacts

import android.Manifest
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.provider.ContactsContract
import android.provider.Settings
import android.telephony.TelephonyManager
import com.getcapacitor.JSArray
import com.getcapacitor.JSObject
import com.getcapacitor.PermissionState
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import com.getcapacitor.annotation.Permission
import com.getcapacitor.annotation.PermissionCallback
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import java.util.Locale

/**
 * Read-only contact lookup for Connect matching.
 *
 * Contacts are returned to the web layer for in-memory hashing only. The web
 * layer sends hashes to the backend and does not persist raw contact records.
 *
 * The content-resolver scan runs off the main thread: a contact book of a few
 * thousand rows blocks long enough to trip an ANR if queried inline.
 */
@CapacitorPlugin(
    name = "HushhContacts",
    permissions = [
        Permission(
            alias = "contacts",
            strings = [Manifest.permission.READ_CONTACTS]
        )
    ]
)
class HushhContactsPlugin : Plugin() {

    private companion object {
        const val DEFAULT_LIMIT = 5000
        const val MAX_LIMIT = 10000
    }

    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    override fun handleOnDestroy() {
        scope.cancel()
        super.handleOnDestroy()
    }

    @PluginMethod
    fun getPermissionState(call: PluginCall) {
        call.resolve(permissionPayload())
    }

    @PluginMethod
    fun requestPermission(call: PluginCall) {
        if (getPermissionState("contacts") == PermissionState.GRANTED) {
            call.resolve(permissionPayload())
            return
        }
        requestPermissionForAlias("contacts", call, "contactsPermissionStateCallback")
    }

    @PluginMethod
    fun openAppSettings(call: PluginCall) {
        try {
            val intent = Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS).apply {
                data = Uri.parse("package:${context.packageName}")
                addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            }
            context.startActivity(intent)
            call.resolve(JSObject().put("opened", true))
        } catch (error: Exception) {
            call.reject("Could not open app settings: ${error.message}")
        }
    }

    @PluginMethod
    fun readContacts(call: PluginCall) {
        if (getPermissionState("contacts") != PermissionState.GRANTED) {
            requestPermissionForAlias("contacts", call, "contactsPermissionCallback")
            return
        }
        resolveContacts(call)
    }

    @PermissionCallback
    private fun contactsPermissionStateCallback(call: PluginCall) {
        call.resolve(permissionPayload())
    }

    @PermissionCallback
    private fun contactsPermissionCallback(call: PluginCall) {
        if (getPermissionState("contacts") != PermissionState.GRANTED) {
            call.reject("Contacts permission was not granted.")
            return
        }
        resolveContacts(call)
    }

    private fun permissionPayload(): JSObject {
        // Capacitor tracks whether the OS will still prompt, which
        // checkSelfPermission cannot distinguish. Without that split the web
        // layer cannot tell "never asked" from "denied for good", and offers
        // a prompt that silently never appears.
        val state = when (getPermissionState("contacts")) {
            PermissionState.GRANTED -> "granted"
            PermissionState.DENIED -> "denied"
            else -> "prompt"
        }
        return JSObject()
            .put("state", state)
            .put("sourcePlatform", "android")
    }

    /** Home number-plan region for bare national numbers. */
    private fun deviceRegion(): String? {
        val telephony = runCatching {
            context.getSystemService(Context.TELEPHONY_SERVICE) as? TelephonyManager
        }.getOrNull()
        // Return only home number-plan evidence here. The shared web layer ranks an
        // Android `defaultRegion` above the signed-in account's verified phone,
        // so falling back to the UI locale here would silently reinterpret an
        // Indian national number as US (or vice versa) on Wi-Fi-only devices.
        // A serving-network country is also unsafe while roaming. When the SIM
        // provides no region, return null and let
        // the shared resolver use the verified account phone before locale.
        val candidates = listOf(runCatching { telephony?.simCountryIso }.getOrNull())
        return candidates
            .firstOrNull { !it.isNullOrBlank() && it.length == 2 }
            ?.uppercase(Locale.US)
    }

    private fun resolveContacts(call: PluginCall) {
        val limit = (call.getInt("limit", DEFAULT_LIMIT) ?: DEFAULT_LIMIT)
            .coerceIn(1, MAX_LIMIT)

        scope.launch {
            try {
                val payload = queryContacts(limit)
                call.resolve(payload)
            } catch (error: Exception) {
                call.reject("Contacts could not be read: ${error.message}")
            }
        }
    }

    private fun queryContacts(limit: Int): JSObject {
        val contactsById = LinkedHashMap<String, ContactAccumulator>()
        // The cursor yields one row per phone number, so a contact with three
        // numbers appears three times. Counting rows would inflate the total
        // for every contact past the limit; count distinct contacts instead.
        val seenContactIds = HashSet<String>()
        val projection = arrayOf(
            ContactsContract.CommonDataKinds.Phone.CONTACT_ID,
            ContactsContract.CommonDataKinds.Phone.DISPLAY_NAME,
            ContactsContract.CommonDataKinds.Phone.NUMBER,
            ContactsContract.CommonDataKinds.Phone.NORMALIZED_NUMBER
        )

        context.contentResolver.query(
            ContactsContract.CommonDataKinds.Phone.CONTENT_URI,
            projection,
            null,
            null,
            // Ordering by contact id uses the primary index; DISPLAY_NAME forces
            // a full sort and makes truncation alphabetical, which silently hid
            // everyone past the letter the cut landed on.
            ContactsContract.CommonDataKinds.Phone.CONTACT_ID + " ASC"
        )?.use { cursor ->
            val idIndex = cursor.getColumnIndex(ContactsContract.CommonDataKinds.Phone.CONTACT_ID)
            val nameIndex = cursor.getColumnIndex(ContactsContract.CommonDataKinds.Phone.DISPLAY_NAME)
            val phoneIndex = cursor.getColumnIndex(ContactsContract.CommonDataKinds.Phone.NUMBER)
            val normalizedIndex =
                cursor.getColumnIndex(ContactsContract.CommonDataKinds.Phone.NORMALIZED_NUMBER)

            while (cursor.moveToNext()) {
                val id = cursor.getString(idIndex) ?: continue
                val number = cursor.getString(phoneIndex)?.trim().orEmpty()
                val normalized = if (normalizedIndex >= 0) {
                    cursor.getString(normalizedIndex)?.trim().orEmpty()
                } else {
                    ""
                }
                if (number.isEmpty() && normalized.isEmpty()) continue

                seenContactIds.add(id)
                if (!contactsById.containsKey(id) && contactsById.size >= limit) {
                    continue
                }

                val name = cursor.getString(nameIndex)?.trim().orEmpty()
                val entry = contactsById.getOrPut(id) { ContactAccumulator(id, name) }
                // Android keeps its own E.164 rendering of each number when it
                // can derive one. It is strictly better than re-deriving from
                // the display string, so it goes first.
                if (normalized.isNotEmpty() && !entry.phoneNumbers.contains(normalized)) {
                    entry.phoneNumbers.add(0, normalized)
                }
                if (number.isNotEmpty() && !entry.phoneNumbers.contains(number)) {
                    entry.phoneNumbers.add(number)
                }
            }
        }

        val contacts = JSArray()
        contactsById.values.forEach { entry ->
            val phoneNumbers = JSArray()
            entry.phoneNumbers.forEach { phoneNumbers.put(it) }
            contacts.put(
                JSObject()
                    .put("id", entry.id)
                    .put("displayName", entry.displayName)
                    .put("phoneNumbers", phoneNumbers)
            )
        }

        return JSObject()
            .put("contacts", contacts)
            .put("sourcePlatform", "android")
            .put("defaultRegion", deviceRegion())
            // Android has no partial-contact mode; access is all or nothing.
            .put("limited", false)
            .put("truncated", seenContactIds.size > contactsById.size)
            .put("totalAvailable", seenContactIds.size)
    }

    private data class ContactAccumulator(
        val id: String,
        val displayName: String,
        val phoneNumbers: MutableList<String> = mutableListOf()
    )
}
